In [118]:
import pandas as pd
import numpy as np

## Reading files

In [119]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [120]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [121]:
wh = pd.concat([wh2018,wh2019])

In [122]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [123]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [124]:
year_group = wh.groupby('year').count()

In [125]:
nansdict ={}
for year in year_group.index:
    nansdict[year] = {}

In [126]:
for year in year_group.index:
    for c in wh.columns[0:-1]:
        nansdict[year][c] = wh.loc[wh['year'] == year][c].isnull().sum() / year_group[c][year]*100


In [127]:
dataframeNullPorcentage = pd.DataFrame(nansdict)

In [128]:
dataframeNullPorcentage

2018      2019
Overall rank                   0.000000  0.000000
Country or region              0.000000  0.000000
Score                          3.311258  9.859155
GDP per capita                 6.122449  1.960784
Social support                10.638298  0.645161
Healthy life expectancy        8.333333  9.859155
Freedom to make life choices   8.333333  7.586207
Generosity                    12.230216  2.631579
Perceptions of corruption      4.697987  4.000000

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [129]:
for year in wh['year'].unique().tolist():
    print(f'Summary of year {year}\n')
    print(wh.loc[wh['year'] == year].describe())
    print('\n')

Summary of year 2018

       Overall rank       Score  GDP per capita  Social support  \
count    156.000000  151.000000      147.000000      141.000000   
mean      78.500000    5.354060        0.892748        1.205915   
std       45.177428    1.109532        0.399153        0.306070   
min        1.000000    2.905000        0.000000        0.000000   
25%       39.750000    4.451500        0.598500        1.047000   
50%       78.500000    5.358000        0.940000        1.258000   
75%      117.250000    6.154000        1.214500        1.462000   
max      156.000000    7.632000        2.096000        1.592000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               144.000000                    144.000000  139.000000   
mean                  0.597576                      0.456514    0.181360   
std                   0.250700                      0.162312    0.099147   
min                   0.000000                      0.000000    0.0000

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [130]:
meanscolumns = wh.groupby('year').mean()

In [131]:
arrayearsfilled =[]

In [132]:
for year in wh['year'].unique().tolist():
    wh.loc[wh['year'] == year] = wh.loc[wh['year'] == year].fillna(meanscolumns.loc[year])	

In [133]:
wh

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0          1.592000                    0.874                         0.681   
1          1.582000                    0.861                         0.686   
2          1.590000                    0.868                         0.683   
3          1.205915                    0.914                         0.677   
4          1.549000                    0.927                         0.660   
..              ...                      ...                           ...   
151        0.711000                    0.614                         0.555   
152        0.885000                    0.499                         0.417   
153        0.517000                    0.361                         0.000   
154        0.000000                    0.105                         0.225   
155        0.575000                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0       0.20200                      0.393  2018  
1       0.28600                      0.340  2018  
2       0.28400                      0.408  2018  
3       0.18136                      0.138  2018  
4       0.25600                      0.357  2018  
..          ...                        ...   ...  
151     0.21700                      0.411  2019  
152     0.27600                      0.147  2019  
153     0.15800                      0.025  2019  
154     0.23500                      0.035  2019  
155     0.20200                      0.091  2019  

[312 rows x 10 columns]

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [134]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [135]:
paises = wh['Country or region'].unique().tolist()

In [136]:
diferenciaencontrada = pd.DataFrame(index = wh['Country or region'])

In [137]:
for i in diferenciaencontrada.index:
    pass
    #diferenciaencontrada =  wh.loc[wh['Country or region'] == 'Switzerland' , : ]